helion/language/ref_tile.py

```diff
@@ -0,0 +1,165 @@
+
+        if func is torch.Tensor.__setitem__:
+            return cls._handle_setitem(func, args, kwargs)
+
+        if func is torch.Tensor.__format__:
+            return repr(args[0])
+
+        raise exc.IncorrectTileUsage(func)
+
+    @classmethod
+    def _handle_getitem(
+        cls,
+        func: Callable[..., object],
+        args: tuple[object, ...],
+        kwargs: dict[str, object] | None,
+    ) -> object:
+        """Handle tensor[index] operations."""
+        tensor, index = args
+        assert isinstance(tensor, torch.Tensor)
+
+        if isinstance(index, RefTile):
+            return tensor[index._slice]
+
+        if isinstance(index, tuple):
+            new_index = cls._convert_tile_indices_to_slices(index)
+            return tensor[tuple(new_index)]  # pyright: ignore[reportArgumentType]
+
+        # Non-tile index in ref mode
+        return tensor[index]  # pyright: ignore[reportArgumentType]
+
+    @classmethod
+    def _handle_setitem(
+        cls,
+        func: Callable[..., object],
+        args: tuple[object, ...],
+        kwargs: dict[str, object] | None,
+    ) -> object:
+        """Handle tensor[index] = value operations."""
+        tensor, index, value = args
+        assert isinstance(tensor, torch.Tensor)
+        assert isinstance(value, (int, float, bool, torch.Tensor))
+
+        if isinstance(index, RefTile):
+            tensor[index._slice] = value
+            return None
+
+        if isinstance(index, tuple):
+            new_index = cls._convert_tile_indices_to_slices(index)
+            tensor[tuple(new_index)] = value  # pyright: ignore[reportArgumentType]
+            return None
+
+        # Non-tile index in ref mode
+        tensor[index] = value  # pyright: ignore[reportArgumentType]
+        return None
+
+    @classmethod
+    def _convert_tile_indices_to_slices(
+        cls, indices: tuple[object, ...]
+    ) -> list[object]:
+        """Convert RefTile objects in a tuple of indices to slices."""
+        new_index = []
+        for idx in indices:
+            if isinstance(idx, RefTile):
+                new_index.append(idx._slice)
+            else:
+                new_index.append(idx)
+        return new_index
+
+    def __repr__(self, tensor_contents: None = None) -> str:  # pyright: ignore[reportIncompatibleMethodOverride]
+        return f"RefTile({self._slice!r})"
+
+    @property
+    def index(self) -> torch.Tensor:
+        env = CompileEnvironment.current()
+        return torch.arange(self.start, self.stop, dtype=torch.int32, device=env.device)
+
+    @property
+    def begin(self) -> int:
+        return self._slice.start
+
+    @property
+    def end(self) -> int:
+        return self._slice.stop
+
+    @property
+    def start(self) -> int:
+        return self._slice.start
+
+    @property
+    def stop(self) -> int:
+        return self._slice.stop
+
+    @property
+    def step(self) -> int:
+        return self._slice.step
```

Comment:
Could we share code with the existing Tile class?  None of these will work the `hl.*` versions.

Also I think some of these don't exist in the tile class.  Where did they come from?